### PR TITLE
HSL DateTime: codegen DateTime\Zone enum

### PR DESCRIPTION
**Pull request by jjergus**
Includes the codegen logic (`bin/codegen-timezones.php`) and the generated file. The docblock explains the rationale for what is and isn't included in the enum.

I'm not sure if we want to commit the codegen logic, but it probably makes sense to keep it somewhere so that the enum can be regenerated e.g. when HHVM is built in the future with a newer version of tzdata.

### CODE REVIEW

File: bin/codegen-timezones.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    invariant(
+      Regex\matches($id, re"#^[A-Za-z]+/[A-Za-z_/\\-]+\$#"),
+      'Unexpected timezone identifier format: %s',
+      $id,
+    );
+    $enum_values[Regex\replace($id, re"#[/\\-]#", '_') |> Str\uppercase($$)] =
+      $id;
+  }
+
+  //\print_r($enum_values);
```
Comment (fredemmott) [inline bin/codegen-timezones.php:111]: delete

File: bin/codegen-timezones.php [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+- all supported tzdata timezones like "America/Los_Angeles"
+- all UTC offsets that at least one supported timezone uses for either its
+  winter time or its summer time
+
+We intentionally don't include common aliases like PST (Pacific Standard
+Time) because these can be confusing and ambiguous, e.g. the meaning of PST
+is unclear when dealing with dates that fall in PDT (Pacific Daylight Time)
+range. Instead, use the standard tzdata timezone like "America/Los_Angeles"
+which correctly resolves to either PST or PDT for any specified date; or
+you can use an explicit offset like UTC-08:00 which is always unambiguous.
```
Comment (fredemmott) [inline bin/codegen-timezones.php:41]: This is reasonable (and best practice), but we probably need these somewhere; separate enum?

**Review by fredemmott — APPROVED** [review]
Do you plan on using this as a parameter/return type etc?

If this is just for convenience, no problem - but if it's planned to be used as a constraint, outside of Facebook, HHVM reads the system tzdata, so new timezones can be added just by running `apt-get upgrade` - arbitrary strings should probably be supported.

For a little more safety, perhaps a type alias that can be trivially created from either the enum or a string, but the string version gets a scary UNSAFE name?

**Comment by jjergus** [comment]
I think it's better to start more restrictive, then we can relax the restrictions later if anyone needs it (it's much harder to go the other way), so for now I would rather avoid supporting arbitrary strings as well as any confusing/ambiguous/deprecated aliases.

File: bin/codegen-timezones.php [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+namespace HH\Lib\Experimental\DateTime\_Private;
+
+use namespace HH\Lib\{Keyset, Regex, Str};
+use type Facebook\HackCodegen\{
+  CodegenFileType,
+  HackBuilderValues,
+  HackCodegenConfig,
+  HackCodegenFactory,
+};
```
[the remark this answers is not in the document]
Comment (jjergus) [reply bin/codegen-timezones.php:20]: Is it a problem even if it's just a "require-dev" dependency?

[the remark this answers is not in the document]
Comment (fredemmott) [reply bin/codegen-timezones.php:20]: I think so: if this becomes built-in, we need both FB and non-FB contributors to be comfortable running this

**Pull request merged**